posts/2018-11-28-http4s-error-handling-mtl-2.md

```diff
@@ -37,25 +37,19 @@ I recently gave a 20 minutes talk on `classy optics` at the unconference of [Sca
 
 Something that made me cringe and that a few of my colleagues at work were not happy with was that the algebras had no association with the error type defined in `HttpErrorHandler[F, E]` so the type-safety was down to the programmer's discipline and in this case the compiler was not able to do much.
 
-When working with `EitherT[F, E, A]` or a bifunctor `IO[E, A]` we have a clear error type whereas by just relying on a single `F[A]` with a `MonadError[F, Throwable]` instance we lose this property. However, most of the time we only care about mapping a few business errors to Http responses if working on a REST API.
+When working with `EitherT[F, E, A]` or a bifunctor `IO[E, A]` we have a clear error type whereas by just relying on a single `F[A]` with a `MonadError[F, Throwable]` instance we lose this property. There are a few issues with the first though:
 
-#### Data types with error handling capabilities
+- It has a "double error channel", meaning that can report errors via `Left` or via its effect type `IO`.
+- As any other Monad Transformer in Scala, introduces a performance overhead due to the extra `flatMap` calls and extra allocations.
+- Code becomes cumbersome as we need to lift effects and pure `Either` values into the transformer stack.
 
-As of now I still prefer the latter for a number of reasons. Here's a comparison table of all the different approaches I can think of:
+The `IO[E, A]` model is naturally a better approach but I found out polymorphic code is more cumbersome than working with `F[A]`.
```

Comment:
It is more cumbersome with final tagless right now, although the main reason is that cats-effect doesn't have bifunctor type classes... yet. 🤞

Reply:
Linked ;)